### PR TITLE
New functionality and bugfixes

### PR DESCRIPTION
**Pull request by tandy-1000**
Adds support for some endpoints in [Rooms](https://matrix.org/docs/spec/client_server/r0.6.1#post-matrix-client-r0-createroom).
Bug fix for endutils and register endpoints.

---

Still WIP! This is my todo list :)

 - [x] Implement relevant APIs for matrix-nim-sdk
    - [x] `/_matrix/client/r0/rooms/$roomId/join`
    - [x] `/_matrix/client/r0/register`
    - [x] `/_matrix/client/r0/createRoom`
    - [x] `/_matrix/client/r0/sync`
    - [x] `/_matrix/client/r0/rooms/{roomId}/send/{eventType}/{txnId}`
    - [x] `/_matrix/client/r0/rooms/$roomId/state` - for room name, etc

### CODE REVIEW

**Comment by dylhack** [comment]
Thank you for this, let me check it out.

**Comment by tandy-1000** [comment]
> Thank you for this, let me check it out.

Still WIP! This is my todo list :)

 - [ ] Implement relevant APIs for matrix-nim-sdk
    - [ ] `/_matrix/client/r0/rooms/$roomId/join`
    - [x] `/_matrix/client/r0/register`
    - [x] `/_matrix/client/r0/createRoom`
    - [ ] `/_matrix/client/r0/sync`
    - [ ] `/_matrix/client/r0/rooms/{roomId}/send/{eventType}/{txnId}`
    - [ ] `/_matrix/client/r0/rooms/$roomId/state` - for room name, etc

File: src/matrix/clientserver/rooms.nim [hunk 1]
```diff
@@ -0,0 +1,3 @@
+import rooms/[creation, aliases]
+
+export creation
```
Comment (dylhack) [inline src/matrix/clientserver/rooms.nim:3]: Do we need to export aliases as well?

File: src/matrix/core/endutils.nim [hunk 2]
```diff
@@ -39,7 +39,7 @@ func addQuery*(e: var Endpoint, query: openArray[(string, string)]): void =
 func build*(
   endpoint: EndpointDraft,
   homeserver: Uri,
-  params: varargs[(string, string)] = []): Endpoint =
+  urlParams, queryParams: varargs[(string, string)] = []): Endpoint =
```
Comment (dylhack) [inline src/matrix/core/endutils.nim:42]: Do the other functions work normally with this change?

**Comment by dylhack** [comment]
> > Thank you for this, let me check it out.
> 
> Still WIP! This is my todo list :)
> 
> * [ ]  Implement relevant APIs for matrix-nim-sdk
>   
>   * [ ]  `/_matrix/client/r0/rooms/$roomId/join`
>   * [x]  `/_matrix/client/r0/register`
>   * [x]  `/_matrix/client/r0/createRoom`
>   * [ ]  `/_matrix/client/r0/sync`
>   * [ ]  `/_matrix/client/r0/rooms/{roomId}/send/{eventType}/{txnId}`
>   * [ ]  `/_matrix/client/r0/rooms/$roomId/state` - for room name, etc

Okay, gotcha! Thank you so much for this. I really appreciate it.

File: src/matrix/clientserver/rooms.nim (hunk 1 above)
Comment (tandy-1000) [reply src/matrix/clientserver/rooms.nim]: yes! missed a spot :)

File: src/matrix/core/endutils.nim (hunk 2 above)
Comment (tandy-1000) [reply src/matrix/core/endutils.nim:42]: kinda, you need to specify whether they are `urlParams` or `queryParams`, but I added a new test for that

**Comment by tandy-1000** [comment]
> > > Thank you for this, let me check it out.
> > 
> > 
> > Still WIP! This is my todo list :)
> > 
> > * [ ]   Implement relevant APIs for matrix-nim-sdk
> >   
> >   * [ ]   `/_matrix/client/r0/rooms/$roomId/join`
> >   * [x]   `/_matrix/client/r0/register`
> >   * [x]   `/_matrix/client/r0/createRoom`
> >   * [ ]   `/_matrix/client/r0/sync`
> >   * [ ]   `/_matrix/client/r0/rooms/{roomId}/send/{eventType}/{txnId}`
> >   * [ ]   `/_matrix/client/r0/rooms/$roomId/state` - for room name, etc
> 
> Okay, gotcha! Thank you so much for this. I really appreciate it.

Jump into the matrix chat, I'm working on a simple client atm :))

File: src/matrix/core/endutils.nim (hunk 2 above)
Comment (dylhack) [reply src/matrix/core/endutils.nim:42]: Okay, I will look at the functions as well to make sure they're all good.

**Comment by dylhack** [comment]
My database corrupted for Synapse in postgres, so I haven't bothered fixing it. I will do that today and see if I can get in the room again.

**Comment by dylhack** [comment]
I might try dendrite this time

**Comment by tandy-1000** [comment]
> My database corrupted for Synapse in postgres, so I haven't bothered fixing it. I will do that today and see if I can get in the room again.

oh no :( hope you can figure it out, good luck with dendrite, conduit seems to be good too

**Comment by dylhack** [comment]
Ooo if conduit has more features accomplished I might try that

**Comment by tandy-1000** [comment]
I think I want to fix CI in this PR as well, so I'll do that before I mark as ready.

**Comment by tandy-1000** [comment]
Nope, I'll do that another day lol.

**Comment by tandy-1000** [comment]
I'm going to refactor to use the new jshttpclient which supports sync/async https://github.com/juancarlospaco/nodejs/pull/4

**Comment by tandy-1000** [comment]
Done, I'm going to test the JS stuff in the browser now.

**Comment by tandy-1000** [comment]
> I'm going to refactor to use the new jshttpclient which supports sync/async [juancarlospaco/nodejs#4](https://github.com/juancarlospaco/nodejs/pull/4)

Finished this, tried to implement it in the browser for my matrix client and got stuck on headers not being defined properly...
XMLHTTPRequest is probably to buggy for this to ever work properly..

File: matrix.nimble [hunk 3]
```diff
@@ -8,7 +8,7 @@ srcDir = "src"
 
 # Dependencies
 requires "nim >= 1.5.1"
-requires "jsony 1.0.5"
+requires "jsony#head"
```
[the remark this answers is not in the document]
Comment (dylhack) [reply matrix.nimble:11]: Use this

```nim
requires "jsony#d0e69bddf83874e15b5c2f52f8b1386ac080b443"
```

[the remark this answers is not in the document]
Comment (dylhack) [reply matrix.nimble:11]: @tandy-1000 Use this

```nim
requires "jsony#d0e69bddf83874e15b5c2f52f8b1386ac080b443"
```

[the remark this answers is not in the document]
Comment (tandy-1000) [reply matrix.nimble]: good idea, thanks :)

**Pull request closed without merging**